consensus/src/chained_bft/liveness/leader_reputation.rs

```diff
@@ -0,0 +1,131 @@
+    fn get_meta_data(&self, window_size: u64) -> &[BlockMetadata];
+}
+
+/// Interface to calculate weights for proposers based on history.
+pub trait ReputationHeuristic {
+    /// Return the weights of all candidates based on the history.
+    fn get_weights(&self, candidates: &[Author], history: &[BlockMetadata]) -> Vec<u64>;
+}
+
+/// Simple heuristic: If candidate appear in the history, it's assigned active_weight otherwise inactive weight.
+pub struct SimpleHeuristic {
+    active_weight: u64,
+    inactive_weight: u64,
+}
+
+impl SimpleHeuristic {
+    pub fn new(active_weight: u64, inactive_weight: u64) -> Self {
+        Self {
+            active_weight,
+            inactive_weight,
+        }
+    }
+}
+
+impl ReputationHeuristic for SimpleHeuristic {
+    fn get_weights(&self, candidates: &[Author], history: &[BlockMetadata]) -> Vec<u64> {
+        let set = history.iter().fold(HashSet::new(), |mut set, meta| {
+            set.insert(meta.proposer());
+            set.extend(meta.voters().into_iter());
+            set
+        });
+        candidates
+            .iter()
+            .map(|author| {
+                if set.contains(&author) {
+                    self.active_weight
+                } else {
+                    self.inactive_weight
+                }
+            })
+            .collect()
+    }
+}
+
+/// Committed history based proposer election implementation that could help bias towards
+/// successful leaders to help improve performance.
+pub struct LeaderReputation<T> {
+    proposers: Vec<Author>,
+    backend: Arc<dyn MetadataBackend>,
+    window_size: u64,
+    heuristic: Box<dyn ReputationHeuristic>,
+    phantom: PhantomData<T>,
+}
+
+#[allow(dead_code)]
+impl<T> LeaderReputation<T> {
+    pub fn new(
+        proposers: Vec<Author>,
+        backend: Arc<dyn MetadataBackend>,
+        window_size: u64,
+        heuristic: Box<dyn ReputationHeuristic>,
+    ) -> Self {
+        Self {
+            proposers,
+            backend,
+            window_size,
+            heuristic,
+            phantom: PhantomData,
+        }
+    }
+
+    pub fn get_rng(&self, round: Round) -> StdRng {
+        let mut seed = [0u8; 32];
+        seed[0..8].clone_from_slice(&round.to_be_bytes());
+        StdRng::from_seed(seed)
+    }
+}
+
+impl<T> ProposerElection<T> for LeaderReputation<T> {
+    fn is_valid_proposer(&self, author: Author, round: Round) -> Option<Author> {
+        if self.get_valid_proposers(round).contains(&author) {
+            Some(author)
+        } else {
+            None
+        }
+    }
+
+    fn get_valid_proposers(&self, round: Round) -> Vec<Author> {
+        let sliding_window = self.backend.get_meta_data(self.window_size);
+        let weighted_index =
+            WeightedIndex::new(self.heuristic.get_weights(&self.proposers, sliding_window))
+                .unwrap();
+        let mut rng = self.get_rng(round);
+        vec![self.proposers[weighted_index.sample(&mut rng)]]
```

Comment:
Why not use the same next() fn as multi_proposer_election?

Reply:
`next` doesn't support weight nor implement rng

Reply:
The downside of using Rng is that any other implementor would need to support that whereas with next it uses SHA3-256, which is used for hashing anyway.

Reply:
Building a simple WeightedIndex implementation shouldn't be too difficult?  Essentially call next() and then index into the vector of weights?

Reply:
I thought I don't have to rebuild the wheel, but yeah it's simple